crypto/init.c

```diff
@@ -521,15 +679,66 @@ int OPENSSL_init_crypto(uint64_t opts, const OPENSSL_INIT_SETTINGS *settings)
-     * return above.
+     * therefore, this code block is below the INIT_BASE_ONLY-conditioned
+     * early return above.
+     *
+     * The OPENSSL_INIT_PROVIDER option is used by providers that are
+     * linked with libcrypto, to enable a functionality similar to doing
+     * atexit(OPENSSL_cleanup), but without actually doing it with atexit().
+     * We do this by using the upcall OSSL_FUNC_core_hook_cleanup(), which
+     * will cause the OPENSSL_cleanup() to be called when the provider is
+     * torn down.
+     * Care is taken to handle the case where libcrypto is shared by multiple
+     * providers, by registering the provider handle.  When the cleanup
+     * function is called, that same handle is unregistered, when when the
+     * register is empty, OPENSSL_cleanup() is called.
+     * Care is also taken to handle the case where the libcrypto that the
+     * provider is linked with is shared with the calling application, by
+     * registering a "provider handle" for it, namely NULL, which ensures
+     * that provider tear down will not cause an OPENSSL_cleanup() call,
+     * and leave it to the OPENSSL_cleanup() call which has been registered
+     * through atexit(), or that the application does explicitly.
      */
     if ((opts & OPENSSL_INIT_NO_ATEXIT) != 0) {
         if (!RUN_ONCE_ALT(&register_atexit, ossl_init_no_register_atexit,
                           ossl_init_register_atexit))
             return 0;
-    } else if (!RUN_ONCE(&register_atexit, ossl_init_register_atexit)) {
-        return 0;
+    } else if ((opts & OPENSSL_INIT_PROVIDER) != 0) {
+        int ret;
+        int (*c_hook_cleanup)(const OSSL_CORE_HANDLE *handle,
+                              void (*cleanup_function)(const OSSL_CORE_HANDLE
+                                                       *handle))
+            = NULL;
+        const OSSL_DISPATCH *upcalls = settings->upcalls;
+
+        /* Settings MUST be given, and the upcalls field MUST be set */
+        if (settings == NULL
+            || settings->upcalls == NULL
+            || settings->handle == NULL)
+            return 0;
+        ret = RUN_ONCE_ALT(&register_atexit,
+                           ossl_init_register_provider_atexit,
+                           ossl_init_register_atexit);
+
+        if (ret <= 0)
+            return 0;
+
+        for (upcalls = settings->upcalls; upcalls->function_id != 0; upcalls++) {
+            switch (upcalls->function_id) {
+            case OSSL_FUNC_CORE_HOOK_CLEANUP:
+                c_hook_cleanup = OSSL_FUNC_core_hook_cleanup(upcalls);
+                break;
+            }
+        }
+
+        if (c_hook_cleanup == NULL
+            || !c_hook_cleanup(settings->handle, ossl_init_do_cleanup))
+            return 0;
+        ossl_init_register_cleanup(settings->handle);
+    } else {
+        if (!RUN_ONCE(&register_atexit, ossl_init_register_atexit))
+            return 0;
+        ossl_init_register_cleanup(NULL);
```

Comment:
Since this is outside of the RUN_ONCE, is there a possible race condition here?

Reply:
Not sure why there would be.  The core provider code runs OPENSSL_init_crypto() before loading anything, and this line gets executed before OPENSSL_INIT_LOAD_CONFIG is processed, so you tell me.